File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/948-letter-case-permutation.java
package main.java.Medium;

import java.util.ArrayList;
import java.util.List;

/**https://leetcode.com/problems/letter-case-permutation/ */
//Given a string s, you can transform every letter individually to be lowercase or uppercase to create another string.
//Return a list of all possible strings we could create. Return the output in any order.
// 
//Example 1:
//Input: s = "a1b2"
//Output: ["a1b2","a1B2","A1b2","A1B2"]
//Example 2:
//Input: s = "3z4"
//Output: ["3z4","3Z4"]
// 
//Constraints:
//	1 <= s.length <= 12
//	s consists of lowercase English letters, uppercase English letters, and digits.
class Solution568 {
    public List<String> letterCasePermutation(String s) {
        List<String> result = new ArrayList();
        dfs(s.toCharArray(), 0, result);
        return result;
    }

    private void dfs(char[] chars, int index, List<String> result) {
        if (index == chars.length) {
            result.add(new String(chars));
            return;
        }
        if (Character.isDigit(chars[index])) {
            dfs(chars, index + 1, result);
            return;
        }
        chars[index] = Character.toLowerCase(chars[index]);
        dfs(chars, index + 1, result);
        chars[index] = Character.toUpperCase(chars[index]);
        dfs(chars, index + 1, result);
    }
}

File: ScanProject/JSAnalysis/SonarJS/Medium/948-letter-case-permutation.js
/**https://leetcode.com/problems/letter-case-permutation/ */
//Given a string s, you can transform every letter individually to be lowercase or uppercase to create another string.
//Return a list of all possible strings we could create. Return the output in any order.


/**
 * @param {string} s
 * @return {string[]}
 */
var letterCasePermutation = function(s) {
    let result = [];
    let temp = "";
    dfs(0, s, temp, result);
    return result;
    
    
};
function dfs(index, s, temp, result){
    if(index == s.length){
        result.push(temp);
        return;
    }
    if(isLetter(s[index])){
        dfs(index + 1, s, temp + s[index].toLowerCase(), result);
        dfs(index + 1, s, temp + s[index].toUpperCase(), result);
    }else{
        dfs(index + 1, s, temp + s[index], result);
    }
}
function isLetter(c){
    return c.toLowerCase() != c.toUpperCase();
}

    